lib/quicly.c

```diff
@@ -408,27 +414,37 @@ static const quicly_transport_parameters_t default_transport_params = {.max_udp_
                                                                        .max_ack_delay = QUICLY_DEFAULT_MAX_ACK_DELAY,
                                                                        .min_ack_delay_usec = UINT64_MAX};
 
-static __thread int64_t now;
-static __thread size_t in_quicly_send_cnt;
+struct st_quicly_lock_now_state_t {
+    uint8_t unlock_on_exit; /* boolean */
+};
 
-static void update_now(quicly_context_t *ctx)
+static void lock_now(quicly_conn_t *conn, struct st_quicly_lock_now_state_t *state)
 {
-    if (in_quicly_send_cnt == 0) {
-        int64_t newval = ctx->now->cb(ctx->now);
-        if (now < newval)
-            now = newval;
+    if (conn->stash.now != 0) {
+        assert(state != NULL && "caller is not reentrant");
+        state->unlock_on_exit = 0;
+    } else {
+        if (state != NULL)
```

Comment:
shouldn't this be an assert?

Reply:
I see this being NULL in create_connection. Why is that the case? So, do I understand this correctly that it's only set when close_connection is called? I think comments might help here.

Reply:
In short, `state` is used for two purposes:
* Indicate if the caller can be entered when one of the four functions (i.e. quicly_accept, quicly_connect, quicly_receive, quicly_send) are running.
* Retains state to be used when exitting from a function that may or may not be called while one of those four functions are running.

We might want to add a comment explaining this.

Long explanation follows.

quicly_accept, quicly_connect, quicly_receive, quicly_send are not reentrant. While a connection is handled by one of these functions, an application callback cannot pass that connection to any of these four functions.

`lock_now(conn, NULL)` and `unlock_now(conn, NULL)` takes care of that.

In contract, applications can call `quicly_close` anytime. It may or may not have been called from an application callback invoked while one of the aforementioned quicly functions was running.

Therefore, when `lock_now` is being called from `quicly_close`, we have to check if it is invoked while one of those quicly functions are running. If one of them is running, we should respect the current time already being available. Otherwise, we should set time by making a call.

When `quicly_close` exits, the cached `now` should remain the same, if it was invoked while one of those quicly functions are running. Otherwise, we should reset `now` as an indicator that there is no quicly function running.

The state object is used to take care of these ifs.